File: binary_search.c
/*
A C implementation of binary search 
from Grokking Algo textbook Ch1
*/


# include <stdio.h>
# include <math.h>

int binary_search(int num_list[], int size, int item){
/* 
    This functions consumes an array of integers, 
    and its size, and the item to find.
    It return the index of item if it is found,
    otherwise return -1
*/
    int low = 0;
    int high = size -1;

    while(low <= high){
        // mid consumes a float, and returns an int
        int mid = floor((low + high) / 2);
        int guess = num_list[mid];
        if (guess == item){
            return mid;
        }
        else{
            if(guess < item){
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }

    }
    return -1;


}

int main(){
    // observe how to define an array here
    int num_list[] = {1, 2, 3, 5, 8, 13};
    
    // need to pass the size of the array to the function
    int answer = binary_search(num_list, 6, 5);
    printf("the index is %d\n", answer); // should be 3

    int answer2 = binary_search(num_list, 6, 21);
    printf("the index is %d\n", answer2); // should be -1

    int answer3 = binary_search(num_list, 6, 13);
    printf("the index is %d\n", answer3); // should be 5


    return 0;
}

File: binary_search.py
'''
A python implementation of binary search slightly different
from Grokking Algo textbook
'''

import math


def binary_search(list: list, item: int) -> int or None:
    '''
    >>> list1 = [1, 2, 3, 5, 8, 13]
    >>> binary_search(list1, 3)
    2
    >>> list1 = [1, 2, 3, 5, 8, 13]
    >>> binary_search(list1, 18) is None
    True
    >>> list1 = [1, 2, 3, 5, 8, 13]
    >>> binary_search(list1, -1) is None
    True
    >>> list1 = [1, 2, 3, 5, 8, 13]
    >>> binary_search(list1, 1)
    0
    >>> list1 = [1, 2, 3, 5, 8, 13]
    >>> binary_search(list1, 13)
    5
    '''

    low = 0
    high = len(list) - 1

    while low <= high:

        # mid should be in the while loop so after high or low gets updated,
        # mid gets updated, too
        mid = math.floor((high + low)/2)

        guess = list[mid]

        if guess == item:
            return mid
        else:
            if guess < item:
                low = mid + 1
            else:
                high = mid - 1
    return None


def main():
    list1 = [1, 2, 3, 5, 8, 13]
    answer = binary_search(list1, 5)  # should be 3
    print(answer)
    list2 = [1, 2, 3, 5, 8, 13]
    answer = binary_search(list2, -5)  # should be None
    print(answer)


if __name__ == "__main__":
    main()
